src/raftstore/store/worker/region.rs

```diff
@@ -448,11 +441,24 @@ impl SnapContext {
         STALE_PEER_PENDING_DELETE_RANGE_GAUGE.set(self.pending_delete_ranges.len() as f64);
 
         let now = time::Instant::now();
-        let mut timeout_ranges = self.pending_delete_ranges.drain_timeout_ranges(now);
-        for (region_id, start_key, end_key) in timeout_ranges.drain(..) {
-            self.cleanup_range(
-                region_id, start_key, end_key, true, /* use_delete_files */
-            );
+        let mut cleaned_range_keys = vec![];
+        {
+            for (region_id, start_key, end_key) in self.pending_delete_ranges.timeout_ranges(now) {
+                let use_delete_files = true;
+                self.cleanup_range(
+                    region_id,
+                    start_key.as_slice(),
+                    end_key.as_slice(),
+                    use_delete_files,
+                );
+                cleaned_range_keys.push(start_key);
+                if now.elapsed() >= CLEANUP_MAX_BLOCK_DURATION {
+                    break;
+                }
+            }
+        }
+        for key in cleaned_range_keys {
+            self.pending_delete_ranges.remove(&key);
```

Comment:
I think remove should return the removed item, so here you can check if the removed item is expected.